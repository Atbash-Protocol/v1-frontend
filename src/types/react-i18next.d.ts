import { resources } from './i18n';
import 'react-i18next';

declare module 'react-i18next' {
    type DefaultResources = typeof resources['en'];
    interface Resources extends DefaultResources {}
}
