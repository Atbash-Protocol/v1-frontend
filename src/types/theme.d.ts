import '@mui/material/styles';

interface Background {
    dark: string;
}

declare module '@mui/material/styles' {
    interface Palette {
        cardBackground: Palette['primary'];
    }
    interface PaletteOptions {
        cardBackground: PaletteOptions['primary'];
    }
}
