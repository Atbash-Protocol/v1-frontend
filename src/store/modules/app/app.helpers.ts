import { Epoch } from './app.types';

export const calculateStakingRewards = (epoch: Epoch | null, circSupply: number) => {
    const stakingReward = epoch ? epoch.distribute : null; // the amount of BASH to distribute in the coming epoch
    const stakingRebase = stakingReward ? stakingReward.toNumber() / (circSupply * Math.pow(10, 9)) : null; // rewardYield rate for this epoch

    return {
        fiveDayRate: stakingRebase ? Math.pow(1 + stakingRebase, 5 * 3) - 1 : null, // 3 epoch/day
        stakingAPY: stakingRebase ? Math.pow(1 + stakingRebase, 365 * 3) - 1 * 100 : null,
        stakingReward,
        stakingRebase,
    };
};
