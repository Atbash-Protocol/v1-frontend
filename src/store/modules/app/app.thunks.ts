import { JsonRpcProvider } from "@ethersproject/providers";
import { createAsyncThunk } from "@reduxjs/toolkit";
import { ethers } from "ethers";
import { ERC20_DECIMALS } from "lib/contracts/contracts";
import { IReduxState } from "store/slices/state.interface";
import { RootState } from "store/store";

export const getBlockchainData = createAsyncThunk("app/blockchain", async (provider: JsonRpcProvider) => {
    const currentBlock = await provider.getBlockNumber();
    const { timestamp } = await provider.getBlock(currentBlock);

    return {
        currentBlock,
        timestamp,
    };
});

export const getCoreMetrics = createAsyncThunk("app/coreMetrics", async (_, { getState }) => {
    const {
        main: {
            contracts: { BASH, SBASH_ADDRESS, INITIAL_PAIR_ADDRESS },
        },
    } = getState() as RootState;

    const totalSupply = (await BASH!.totalSupply()) / 10 ** ERC20_DECIMALS;
    const circSupply = (await SBASH_ADDRESS!.circulatingSupply()) / Math.pow(10, ERC20_DECIMALS);
    const [reserve1, reserve2]: ethers.BigNumber[] = await INITIAL_PAIR_ADDRESS!.getReserves();

    return {
        totalSupply,
        circSupply,
        reserves: reserve2.div(reserve1).div(10 ** 9),
    };
});

export const getStakingMetrics = createAsyncThunk("app/stakingMetrics", async (_, { getState }) => {
    const {
        main: {
            contracts: { STAKING_ADDRESS },
        },
    } = getState() as IReduxState;

    const epoch = await STAKING_ADDRESS!.epoch();
    const secondsToNextEpoch = 0; // Number(await STAKING_ADDRESS.secondsToNextEpoch());
    // const secondsToNextEpoch = Number(await stakingContract.secondsToNextEpoch());

    const index = await STAKING_ADDRESS!.index();

    return {
        epoch,
        index,
        secondsToNextEpoch,
    };
});
