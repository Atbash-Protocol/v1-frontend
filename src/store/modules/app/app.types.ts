import { Contract, ethers } from 'ethers';

export enum ContractEnum {
    BASH_CONTRACT = 'BASH_CONTRACT',
    SBASH_CONTRACT = 'SBASH_CONTRACT',
    WSBASH_CONTRACT = 'WSBASH_CONTRACT',
    DAI_CONTRACT = 'DAI_CONTRACT',
    STAKING_CONTRACT = 'STAKING_CONTRACT',
    STAKING_HELPER_CONTRACT = 'STAKING_HELPER_CONTRACT',
    BASH_DAI_LP_CONTRACT = 'BASH_DAI_LP_CONTRACT',
    REDEEM_CONTRACT = 'REDEEM_CONTRACT',
    ZAPING_CONTRACT = 'ZAPING_CONTRACT',
}

export interface Epoch {
    distribute: ethers.BigNumber;
    endTime: number;
    number: ethers.BigNumber;
}

export interface MainSliceState {
    contracts: { [key in ContractEnum]: Contract | null };
    contractsLoaded: boolean;
    metrics: {
        totalSupply: number | null;
        circSupply: number | null;
        rawCircSupply: ethers.BigNumber | null;
        reserves: ethers.BigNumber | null;
        loading: boolean;
    };
    staking: {
        epoch: Epoch | null;
        secondsToNextEpoch: number | null;
        index: ethers.BigNumber | null;
        loading: boolean;
    };
    blockchain: {
        currentBlock: number | null;
        timestamp: number | null;
        loading: boolean;
    };
    errorEncountered: boolean; // Flag to prevent page loading
    loading: boolean;
}
