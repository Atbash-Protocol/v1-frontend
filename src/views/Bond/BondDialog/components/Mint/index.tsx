import { Box, Divider } from '@mui/material';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';

import Loader from 'components/Loader';
import MemoInlineMetric from 'components/Metrics/InlineMetric';
import { theme } from 'constants/theme';
import { formatUSD } from 'helpers/price-units';
import { selectBondItemMetrics, selectBondMintingMetrics, selectBondsReady } from 'store/modules/bonds/bonds.selector';
import { BondMetrics } from 'store/modules/bonds/bonds.types';
import { RootState } from 'store/store';

import BondPurchase from '../BondPurchase';

const Mint = ({ bondID, metrics }: { bondID: string; metrics: BondMetrics }) => {
    const { t } = useTranslation();
    const { balance, quote, maxBondPrice, vestingTerm, bondDiscount } = selectBondMintingMetrics(metrics);

    const bondMetrics = [
        { value: balance, metricKey: t('bond:YourBalance') },
        { value: quote, metricKey: t('bond:YouWillGet') },
        { value: maxBondPrice ? formatUSD(maxBondPrice, 2).toString() : null, metricKey: t('bond:MaxYouCanBuy') },
        { metricKey: t('bond:Discount'), value: bondDiscount },
        { metricKey: t('bond:VestingTerm'), value: t('common:day', { count: vestingTerm || 0 }) },
        { metricKey: t('bond:MinimumPurchase'), value: '0.01 BASH' },
    ].map(({ value, metricKey }, index) => <MemoInlineMetric {...{ value, metricKey }} key={index} />);

    return (
        <Box>
            <BondPurchase bondID={bondID} />
            <Divider variant="fullWidth" textAlign="center" sx={{ borderColor: theme.palette.primary.light, marginBottom: theme.spacing(2) }} />
            <Box>{bondMetrics}</Box>
        </Box>
    );
};

const MintLoader = ({ bondID }: { bondID: string }) => {
    const bondsReady = useSelector(selectBondsReady);
    const metrics = useSelector((state: RootState) => selectBondItemMetrics(state, bondID));

    if (!bondsReady) return <Loader />;

    return <Mint bondID={bondID} metrics={metrics} />;
};

export default MintLoader;
