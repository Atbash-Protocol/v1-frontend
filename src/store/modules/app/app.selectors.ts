import Decimal from 'decimal.js';
import { BigNumber, utils } from 'ethers';
import { pick } from 'lodash';
import { createSelector } from 'reselect';

import { formatUSD } from 'helpers/price-units';
import { IReduxState } from 'store/slices/state.interface';
import { RootState } from 'store/store';

import { calculateStakingRewards } from './app.helpers';

export const selectFormattedReservePrice = (state: RootState): string | null => {
    const { reserves } = state.main.metrics;
    const { dai } = state.markets.markets;

    if (!reserves || !dai) return null;

    const reservePrice = Number(utils.formatUnits(reserves.toString(), 'gwei')) * dai;

    return formatUSD(reservePrice, 2);
};

export const selectReserve = (state: RootState): Decimal => new Decimal(state.main.metrics.reserves?.toString() ?? 0);
export const selectReserveLoading = (state: RootState): boolean => state.main.metrics.reserves === null;

export const selectBlockchainLoading = (state: RootState) => state.main.blockchain.loading === true;
export const selectMetricsLoading = (state: RootState) => state.main.metrics.loading === true;
export const selectStakingLoading = (state: RootState) => state.main.staking.loading === true;
export const selectContractsLoading = (state: RootState) => Object.values(state.main.contracts).some(contract => contract === null);

export const useContractLoaded = (state: RootState): boolean => {
    const { contracts } = state.main;

    return Object.values(contracts).every(contract => contract !== null);
};

export const selectAppLoading = createSelector(
    [selectReserveLoading, selectBlockchainLoading, selectMetricsLoading, selectBlockchainLoading, selectContractsLoading],
    (...conditions) => {
        return conditions.some(condition => condition === true);
    },
);

export const useNextRebase = (state: RootState): number | undefined => {
    const {
        main: {
            staking: { epoch },
        },
    } = state;

    return epoch?.endTime;
};

export const useBlockchainInfos = (state: RootState) => state.main.blockchain;

export const selectCirculatingSupply = (state: RootState) => state.main.metrics.circSupply;

export const selectUserStakingInfos = createSelector(
    [
        (state: IReduxState) => {
            return {
                circSupply: state.main.metrics.circSupply,
                ...pick(state.main.staking, ['epoch', 'index']),
                ...pick(state.account.balances, ['SBASH', 'WSBASH']),
            };
        },
    ],
    ({ index, epoch, circSupply, SBASH, WSBASH }) => {
        let stakingRebase = new Decimal(0);

        const SBASHBalance: Decimal = new Decimal(SBASH.div(10 ** 9).toHexString());
        const WBASHBalance: Decimal = new Decimal(WSBASH.div(10 ** 9).toHexString());

        if (epoch?.distribute && circSupply) stakingRebase = new Decimal(epoch.distribute.toHexString()).div(new Decimal(circSupply).mul(10 ** 9));

        const nextRewardValue = stakingRebase.mul(SBASHBalance);
        const { fiveDayRate } = calculateStakingRewards(epoch, circSupply || 0);

        const effectiveNextRewardValue = nextRewardValue.add(stakingRebase.mul(WBASHBalance).mul((index ?? BigNumber.from(0)).toHexString()));
        const wrappedTokenEquivalent = WBASHBalance.mul(index?.toHexString() || 0);

        return {
            fiveDayRate: `${fiveDayRate && (fiveDayRate * 100).toFixed(4)}  %`,
            stakingRebasePercentage: `${stakingRebase.mul(100).toFixed(2)} %`,
            nextRewardValue: `${nextRewardValue.toFixed(2)} BASH`,
            effectiveNextRewardValue: effectiveNextRewardValue ? `effectiveNextRewardValue wsBASH` : null,
            wrappedTokenEquivalent: `${wrappedTokenEquivalent?.toString() || 0} sBASH`,
            optionalMetrics: WBASHBalance.gt(0),
        };
    },
);
