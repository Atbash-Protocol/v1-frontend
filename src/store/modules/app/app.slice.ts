import { createSlice } from '@reduxjs/toolkit';

import { isActionRejected } from 'store/utils/action';

import { initializeProviderContracts, getBlockchainData, getCoreMetrics, getStakingMetrics } from './app.thunks';
import { ContractEnum, MainSliceState } from './app.types';

// Define the initial state using that type
const initialState: MainSliceState = {
    contracts: {
        [ContractEnum.BASH_CONTRACT]: null,
        [ContractEnum.SBASH_CONTRACT]: null,
        [ContractEnum.DAI_CONTRACT]: null,
        [ContractEnum.BASH_DAI_LP_ADDRESS]: null,
        [ContractEnum.STAKING_CONTRACT]: null,
        [ContractEnum.STAKING_HELPER_ADDRESS]: null,
        [ContractEnum.WSBASH_ADDRESS]: null,
        [ContractEnum.ZAPIN_ADDRESS]: null,
        [ContractEnum.REDEEM_CONTRACT]: null,
    },
    contractsLoaded: false,

    metrics: {
        circSupply: null,
        rawCircSupply: null,
        totalSupply: null,
        reserves: null,
        loading: true,
    },
    staking: {
        epoch: null,
        secondsToNextEpoch: null,
        index: null,
        loading: true,
    },
    blockchain: {
        timestamp: null,
        currentBlock: null,
        loading: true,
    },
    errorEncountered: false,
    loading: true,
};

export const MainSlice = createSlice({
    name: 'app-main',
    initialState,
    reducers: {},
    extraReducers: builder => {
        builder
            .addCase(initializeProviderContracts.pending, state => {
                return { ...state, contractsLoaded: false };
            })
            .addCase(initializeProviderContracts.fulfilled, (state, action) => {
                return { ...state, contracts: action.payload, contractsLoaded: true };
            })
            .addCase(getBlockchainData.fulfilled, (state, action) => {
                return {
                    ...state,
                    blockchain: {
                        ...action.payload,
                        loading: false,
                    },
                };
            })
            .addCase(getCoreMetrics.fulfilled, (state, action) => {
                return {
                    ...state,
                    metrics: {
                        ...action.payload,
                        loading: false,
                    },
                };
            })
            .addCase(getStakingMetrics.fulfilled, (state, action) => {
                return {
                    ...state,
                    staking: {
                        ...action.payload,
                        loading: false,
                    },
                };
            })
            .addMatcher(isActionRejected, (state, action) => {
                if (action.type.startsWith('app/')) state.errorEncountered = true;
            });
    },
});

export const selectContracts = (state: MainSliceState) => state.contracts;

export default MainSlice.reducer;
