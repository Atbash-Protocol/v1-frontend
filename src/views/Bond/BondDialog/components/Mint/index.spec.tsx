import { render, screen } from '@testing-library/react';
import * as ReactReduxModule from 'react-redux';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { ContractEnum } from 'store/modules/app/app.types';

import Mint from '.';

jest.mock('../BondPurchase', () => () => <> Bond Purchase</>);

const renderComponent = (component: JSX.Element, state: any) => {
    const middlewares = [thunk];
    const mockStore = createMockStore(middlewares);

    return render(<ReactReduxModule.Provider store={mockStore(state)}>{component}</ReactReduxModule.Provider>);
};

describe('Mint', () => {
    describe('when the state is not ready', () => {
        it('returns the loader', () => {
            const state = {
                bonds: {
                    bondMetrics: {},
                    bondInstances: {},
                },
                main: {
                    contracts: {},
                },
            };
            const { container } = renderComponent(<Mint bondID="dai" slippage={0.5} recipientAddress="0x" />, state);

            expect(container.getElementsByClassName('MuiCircularProgress-root')).toHaveLength(1);
        });
    });

    describe('When the state is ready', () => {
        it('returns the Mint component', () => {
            const state = {
                bonds: {
                    bondMetrics: {
                        dai: {},
                    },
                    bondInstances: {},
                },
                main: {
                    contracts: {
                        [ContractEnum.BASH_CONTRACT]: jest.fn(),
                    },
                },
            };
            const { container } = renderComponent(<Mint bondID="dai" slippage={0.5} recipientAddress="0x" />, state);

            expect(container).toMatchSnapshot();
        });
    });
});
