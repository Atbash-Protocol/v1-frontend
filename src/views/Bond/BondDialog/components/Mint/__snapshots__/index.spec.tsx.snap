// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mint When the state is ready returns the Mint component 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
     Bond Purchase
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-fimgk4-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:YourBalance
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-15tfpcm-MuiSkeleton-root"
          />
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:YouWillGet
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-15tfpcm-MuiSkeleton-root"
          />
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:MaxYouCanBuy
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-15tfpcm-MuiSkeleton-root"
          />
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:Discount
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-15tfpcm-MuiSkeleton-root"
          />
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:VestingTerm
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          common:day
        </p>
      </div>
      <div
        class="MuiBox-root css-1z137pa"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
        >
          bond:MinimumPurchase
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
        >
          0.01 BASH
        </p>
      </div>
    </div>
  </div>
</div>
`;
